src/analysis/souffle/sensitivity_analysis.dl

```diff
@@ -0,0 +1,142 @@
+      operationHasResult(operation, ingress_point), 
+      operationHasAttribute(operation, [_, $StringAttributePayload(_)]).
+
+// Helper functions to flatten input/operand lists. 
+.decl intermediateFlatten(list:OperandList, sublist:OperandList, list_element:AccessPath)
+
+intermediateFlatten(list, sublist, list_element) :- isOperation(operation), 
+      operationHasOperandList(operation, list),
+      operationHasOperandListSuffix(operation, [list_element, sublist]).
+
+.decl flatten(list:OperandList, list_element:AccessPath)
+flatten(list, list_element) :- intermediateFlatten(list, _, list_element).
+
+// Flattening each individual input per value. 
+.decl isSqlInput(operation: Operation, sql_input: AccessPath)
+
+isSqlInput(operation, sql_input) :- isSqlOperation(operation), 
+      operationHasOperandList(operation, input_list), 
+      flatten(input_list, sql_input). 
+
+// Keeping track of sensitivity score per each input at index. 
+// This is used to calculate the maximum input sensitivity. 
+// It's decomposed into base and inductive cases because recursive 
+// aggregates are not supported yet for the compiler case in souffle
+.decl sqlOperationHasSensitivityScoreAtIndex(operation: Operation, sensitivity_score: SensitvityScore, index: number) 
+
+// Base cases (index 0).
+// Sesnitivity score for initial inputs is 1. 
+sqlOperationHasSensitivityScoreAtIndex(operation, 1, 0) :- isSqlOperation(operation), 
+      operationHasOperandAtIndex(operation, sql_input, 0),
+      isInitialInput(sql_input).
+
+// For all other inputs, we get the final sensitivity score. 
+sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, 0) :- isSqlOperation(operation), 
+      operationHasOperandAtIndex(operation, sql_input, 0),
+      sqlValueSensitivity(sql_input, sensitivity_score).      
+
+// Inductive cases.  
+// We keep track at the score at each index, and ensure that the sensitivity score at index is only 
+// updated if the input at index has sensitivity score greater than the previous sensitivity score.
+// Otherwise, the prev_index sensitivity score is copied at index. 
+sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, prev_index+1) :- 
+      isSqlOperation(operation), 
+      operationHasOperandAtIndex(operation, sql_input, prev_index+1),
+      isSqlInput(operation, sql_input), 
+      ( 
+            sqlValueSensitivity(sql_input, input_sensitivity_score), 
+            sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, prev_index), 
+            sensitivity_score >= input_sensitivity_score; // ensure the prev_index sensitivity_score is higher than that of the input 
+            sqlValueSensitivity(sql_input, sensitivity_score), 
+            sqlOperationHasSensitivityScoreAtIndex(operation, prev_sensitivity_score, prev_index), 
+            sensitivity_score >= prev_sensitivity_score // ensure the input sensitvity_score is higher than that of prev_index 
+      ). 
```

Comment:
Is this equivalent to the following?

```
sqlOperationHasSensitivityScoreAtIndex(operation, max(prev_sensitivity_score, input_sensitivity_score), prev_index + 1) :-
  isSqlOperation(operation),
  operationHasOperandAtIndex(operation, sql_input, prev_index + 1),
  isSqlInput(operation, sql_input),
  sqlValueSensitivity(sql_input, input_sensitivity_score),
  sqlOperationHasSensitivityScoreAtIndex(operation, prev_sensitivity_score, prev_index).
```

If so, that might be a bit easier to follow.

Reply:
Yes! Wow, this is much better. I didn't know you could use max in the head of a rule, but now that I think about it, it makes a lot of sense. Thank you, Mark. Pushing another commit now that changes the code.